File: solutions/range-sum-of-bst/range-sum-of-bst.c
// @Title: Range Sum of BST
// @Author: colinjxc
// @Date: 2022-02-05T03:07:09+08:00
// @URL: https://leetcode-cn.com/problems/range-sum-of-bst


/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */


int rangeSumBST(struct TreeNode* root, int low, int high){
    if (root == NULL) return 0;
    if (root->val < low) {
        return rangeSumBST(root->right, low, high);
    } else if (root->val > high) {
        return rangeSumBST(root->left, low, high);
    }
    return root->val + rangeSumBST(root->left, low, high) + rangeSumBST(root->right, low, high);
}

File: solutions/range-sum-of-bst/range-sum-of-bst.go
// @Title: Range Sum of BST
// @Author: colinjxc
// @Date: 2022-02-05T03:26:47+08:00
// @URL: https://leetcode-cn.com/problems/range-sum-of-bst


/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func rangeSumBST(root *TreeNode, low int, high int) int {
    res := 0
    stack := []*TreeNode{}
    for root != nil || len(stack) > 0 {
        for root != nil {
            stack = append(stack, root)
            root = root.Left
        }
        root = stack[len(stack) - 1]
        stack = stack[:len(stack) - 1]
        if (root.Val >= low && root.Val <= high) {
            res += root.Val
        } else if (root.Val > high) {
            // 可以提前退出
            break;
        }
        root = root.Right
    }
    return res
}
